model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanSamlArtifactSessionMappingStoreProvider.java

```diff
@@ -0,0 +1,126 @@
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.sessions.infinispan;
+
+import org.infinispan.client.hotrod.exceptions.HotRodClientException;
+import org.infinispan.commons.api.BasicCache;
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.Time;
+import org.keycloak.models.AuthenticatedClientSessionModel;
+import org.keycloak.models.SamlArtifactSessionMappingModel;
+import org.keycloak.models.SamlArtifactSessionMappingStoreProvider;
+import org.keycloak.models.sessions.infinispan.entities.ActionTokenValueEntity;
+import org.keycloak.models.sessions.infinispan.util.InfinispanUtil;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+
+/**
+ * @author mhajas
+ */
+public class InfinispanSamlArtifactSessionMappingStoreProvider implements SamlArtifactSessionMappingStoreProvider {
+
+    public static final Logger logger = Logger.getLogger(InfinispanSamlArtifactSessionMappingStoreProvider.class);
+
+    private static final String USER_SESSION_ID = "user.session.id";
+    private static final String CLIENT_SESSION_ID = "client.session.id";
+
+    private final Supplier<BasicCache<UUID, ActionTokenValueEntity>> codeCache;
+
+    public InfinispanSamlArtifactSessionMappingStoreProvider(Supplier<BasicCache<UUID, ActionTokenValueEntity>> actionKeyCache) {
+        this.codeCache = actionKeyCache;
+    }
+
+    private Map<String, String> serializedMapping(AuthenticatedClientSessionModel clientSessionModel) {
+        Map<String, String> serialized = new HashMap<>(2);
+        serialized.put(USER_SESSION_ID, clientSessionModel.getUserSession().getId());
+        serialized.put(CLIENT_SESSION_ID, clientSessionModel.getClient().getId());
+
+        return serialized;
+    }
+
+    @Override
+    public void put(String artifact, int lifespanSeconds, AuthenticatedClientSessionModel clientSessionModel) {
+        ActionTokenValueEntity tokenValue = new ActionTokenValueEntity(serializedMapping(clientSessionModel));
+
+        try {
+            BasicCache<UUID, ActionTokenValueEntity> cache = codeCache.get();
+            long lifespanMs = InfinispanUtil.toHotrodTimeMs(cache, Time.toMillis(lifespanSeconds));
+            cache.put(UUID.nameUUIDFromBytes(artifact.getBytes()), tokenValue, lifespanMs, TimeUnit.MILLISECONDS);
```

Comment:
This would require that UTF-8 representation of `artifact` conforms to _name UUID_ format which is not correct assumption. I believe `artifact` needs to be put there in full
```suggestion
            cache.put(artifact, tokenValue, lifespanMs, TimeUnit.MILLISECONDS);
```

Reply:
`Artifact` doesn't need to conform to UUID format because `nameUUIDFromBytes` is doing `md5` in the background which hash it to the correct format. 

I didn't know that we can actually change parameters of BasicCache just like that, but I changed it to `String, String[]` and it works. 

Do you think it would be a good idea to stick with `nameUUIDFromBytes` so that we save some space in the cache? Currently, each default artifact has 44 Bytes. Using `nameUUIDFromBytes` we can shrink it to 8 Bytes.

Reply:
I see. This is actually a great idea, I didn't realize `nameUUIDFromBytes` could be used like this. +1 for using the `nameUUIDFromBytes` then, though then I'd suggest storing the original artifact as one of the values in the `String` array and check it upon retrieval to prevent collision attacks.
Also ensure that the `artifact.getBytes()` uses the same (UTF-8?) encoding to prevent potential environmental differences.